File: backend/app/schemas/application.py
"""Request/response contracts for the applications endpoints."""

from datetime import datetime
from typing import List, Optional

from pydantic import BaseModel, Field, field_validator

from app.models.application import ApplicationStatus, WorkplaceType


class ApplicationCreate(BaseModel):
    company_name: str = Field(min_length=2, max_length=120)
    job_title: str = Field(min_length=2, max_length=160)
    status: ApplicationStatus = ApplicationStatus.saved
    job_url: Optional[str] = Field(default=None, max_length=2048)
    location: Optional[str] = Field(default=None, max_length=200)
    workplace_type: Optional[WorkplaceType] = None
    salary: Optional[str] = Field(default=None, max_length=100)
    job_description: Optional[str] = Field(default=None, max_length=20000)
    notes: Optional[str] = Field(default=None, max_length=5000)
    applied_at: Optional[datetime] = None
    next_action_at: Optional[datetime] = None

    @field_validator("company_name", "job_title")
    @classmethod
    def strip_required(cls, value: str) -> str:
        return value.strip()

    @field_validator("job_url")
    @classmethod
    def validate_job_url(cls, value: Optional[str]) -> Optional[str]:
        if not value:
            return None
        value = value.strip()
        if not (value.startswith("http://") or value.startswith("https://")):
            raise ValueError("job_url must start with http:// or https://")
        return value


class ApplicationUpdate(BaseModel):
    """All fields optional -- PATCH semantics per architecture.md (only
    fields the client actually sent are changed)."""

    company_name: Optional[str] = Field(default=None, min_length=2, max_length=120)
    job_title: Optional[str] = Field(default=None, min_length=2, max_length=160)
    status: Optional[ApplicationStatus] = None
    job_url: Optional[str] = Field(default=None, max_length=2048)
    location: Optional[str] = Field(default=None, max_length=200)
    workplace_type: Optional[WorkplaceType] = None
    salary: Optional[str] = Field(default=None, max_length=100)
    job_description: Optional[str] = Field(default=None, max_length=20000)
    notes: Optional[str] = Field(default=None, max_length=5000)
    applied_at: Optional[datetime] = None
    next_action_at: Optional[datetime] = None


class ApplicationPublic(BaseModel):
    id: str
    company_name: str
    job_title: str
    status: ApplicationStatus
    job_url: Optional[str] = None
    location: Optional[str] = None
    workplace_type: Optional[WorkplaceType] = None
    salary: Optional[str] = None
    job_description: Optional[str] = None
    notes: Optional[str] = None
    applied_at: Optional[datetime] = None
    next_action_at: Optional[datetime] = None
    archived_at: Optional[datetime] = None
    created_at: datetime
    updated_at: datetime


class ApplicationListResponse(BaseModel):
    items: List[ApplicationPublic]
    total: int
    limit: int
    offset: int


class ApplicationStageCounts(BaseModel):
    saved: int = 0
    applied: int = 0
    screening: int = 0
    interviewing: int = 0
    offer: int = 0
    rejected: int = 0
    withdrawn: int = 0


class ApplicationsSummary(BaseModel):
    total_active: int
    stage_counts: ApplicationStageCounts
    recent: List[ApplicationPublic]

File: frontend/src/types/application.ts
// Mirrors backend/app/schemas/application.py so the API contract only has
// one place it can drift: this file.

export type ApplicationStatus =
  | "saved"
  | "applied"
  | "screening"
  | "interviewing"
  | "offer"
  | "rejected"
  | "withdrawn";

export type WorkplaceType = "onsite" | "remote" | "hybrid";

export const STATUS_ORDER: ApplicationStatus[] = [
  "saved",
  "applied",
  "screening",
  "interviewing",
  "offer",
  "rejected",
  "withdrawn",
];

export const STATUS_LABELS: Record<ApplicationStatus, string> = {
  saved: "Saved",
  applied: "Applied",
  screening: "Screening",
  interviewing: "Interviewing",
  offer: "Offer",
  rejected: "Rejected",
  withdrawn: "Withdrawn",
};

export const WORKPLACE_LABELS: Record<WorkplaceType, string> = {
  onsite: "On-site",
  remote: "Remote",
  hybrid: "Hybrid",
};

export type SortOption = "updated_desc" | "created_desc" | "company_asc" | "title_asc" | "next_action_asc";

export const SORT_OPTIONS_ORDER: SortOption[] = [
  "updated_desc",
  "created_desc",
  "company_asc",
  "title_asc",
  "next_action_asc",
];

export const SORT_LABELS: Record<SortOption, string> = {
  updated_desc: "Recently updated",
  created_desc: "Recently added",
  company_asc: "Company (A–Z)",
  title_asc: "Job title (A–Z)",
  next_action_asc: "Next action date",
};

export interface ApplicationPublic {
  id: string;
  company_name: string;
  job_title: string;
  status: ApplicationStatus;
  job_url: string | null;
  location: string | null;
  workplace_type: WorkplaceType | null;
  salary: string | null;
  job_description: string | null;
  notes: string | null;
  applied_at: string | null;
  next_action_at: string | null;
  archived_at: string | null;
  created_at: string;
  updated_at: string;
}

export interface ApplicationListResponse {
  items: ApplicationPublic[];
  total: number;
  limit: number;
  offset: number;
}

export interface ApplicationStageCounts {
  saved: number;
  applied: number;
  screening: number;
  interviewing: number;
  offer: number;
  rejected: number;
  withdrawn: number;
}

export interface ApplicationsSummary {
  total_active: number;
  stage_counts: ApplicationStageCounts;
  recent: ApplicationPublic[];
}

export type ApplicationEventType = "created" | "status_changed" | "updated" | "archived";

export interface ApplicationEvent {
  id: string;
  application_id: string;
  event_type: ApplicationEventType;
  occurred_at: string;
  metadata: Record<string, unknown> | null;
}

export interface ApplicationEventListResponse {
  items: ApplicationEvent[];
}

export interface ApplicationFormValues {
  company_name: string;
  job_title: string;
  status: ApplicationStatus;
  job_url: string;
  location: string;
  workplace_type: WorkplaceType | "";
  salary: string;
  job_description: string;
  notes: string;
  applied_at: string;
  next_action_at: string;
}

export const EMPTY_APPLICATION_FORM_VALUES: ApplicationFormValues = {
  company_name: "",
  job_title: "",
  status: "saved",
  job_url: "",
  location: "",
  workplace_type: "",
  salary: "",
  job_description: "",
  notes: "",
  applied_at: "",
  next_action_at: "",
};

export interface ApplicationPayload {
  company_name: string;
  job_title: string;
  status?: ApplicationStatus;
  job_url?: string | null;
  location?: string | null;
  workplace_type?: WorkplaceType | null;
  salary?: string | null;
  job_description?: string | null;
  notes?: string | null;
  applied_at?: string | null;
  next_action_at?: string | null;
}
